### PR TITLE
86c0tcetk | Add `fund_pool` function

### PR DESCRIPTION
**Pull request by sayandcode**
## What?
[Clickup Ticket 86c0tcetk](https://app.clickup.com/t/86c0tcetk)
This PR adds a `fund_pool` function, using which anyone can add liquidity to a certain pool and options

## How?
The options are passed in via `remaining_accounts`, and we distribute the `value` among those options. The SOL value is also transferred from the sender to the pool account using a CPI transfer

### CODE REVIEW

File: programs/degen-pools/src/pools.rs [hunk 1]
```diff
@@ -87,6 +89,45 @@ pub fn set_winning_option(ctx: Context<UpdatePool>, winning_option: Pubkey) -> R
     Ok(())
 }
 
+pub fn fund_pool<'c: 'info, 'info>(
+    ctx: Context<'_, '_, 'c, 'info, FundPool<'info>>,
```
Comment (AngSin) [inline programs/degen-pools/src/pools.rs:93]: What is the need for the initial arguments? Can't we follow the same pattern as https://github.com/Degen-Markets/degen-markets-solana/pull/12/files#diff-e4fe02d9ea4178a1c23a95acc10402633301397989504c69116d932a115bfefeR79 ?

File: programs/degen-pools/src/pools.rs [hunk 2]
```diff
@@ -87,6 +89,45 @@ pub fn set_winning_option(ctx: Context<UpdatePool>, winning_option: Pubkey) -> R
...
+        let mut pool_option = match Account::<PoolOption>::try_from(account) {
+            Ok(pool_option) => pool_option,
+            Err(_) => return err!(CustomError::InvalidPoolOptionAccount),
+        };
+        pool_option.value += value_per_option;
+        pool_option.exit(&PROGRAM_ID)?;
+    }
+
+    // Transfer SOL from caller to pool account
+    let cpi_context = CpiContext::new(
```
Comment (AngSin) [inline programs/degen-pools/src/pools.rs:117]: is there a way to use the existing `enter_pool` fn in `entries.rs`?

That will also take away the need to duplicate this check: https://github.com/Degen-Markets/degen-markets-solana/pull/12/files#diff-e4fe02d9ea4178a1c23a95acc10402633301397989504c69116d932a115bfefeR97

File: programs/degen-pools/src/pools.rs [hunk 3]
```diff
@@ -87,6 +89,45 @@ pub fn set_winning_option(ctx: Context<UpdatePool>, winning_option: Pubkey) -> R
...
+    if ctx.remaining_accounts.len() == 0 {
+        return err!(CustomError::NoPoolOptionAccounts);
+    }
+
+    let remaining_accounts = ctx.remaining_accounts;
+    let value_per_option = value / remaining_accounts.len() as u64;
+    for account in remaining_accounts {
+        let mut pool_option = match Account::<PoolOption>::try_from(account) {
+            Ok(pool_option) => pool_option,
+            Err(_) => return err!(CustomError::InvalidPoolOptionAccount),
```
Comment (AngSin) [inline programs/degen-pools/src/pools.rs:110]: We should not just check that it is a valid PoolOption but also check that it belongs to the pool that was passed in the instruction. If we follow [this suggestion](https://github.com/Degen-Markets/degen-markets-solana/pull/12/files#r1829242517), `enter_pool` will do the check for us

File: programs/degen-pools/src/pools.rs [hunk 4]
```diff
@@ -87,6 +89,45 @@ pub fn set_winning_option(ctx: Context<UpdatePool>, winning_option: Pubkey) -> R
...
+        };
+        pool_option.value += value_per_option;
+        pool_option.exit(&PROGRAM_ID)?;
```
Comment (AngSin) [inline programs/degen-pools/src/pools.rs:113]: What does this do?

File: programs/degen-pools/src/pools.rs (hunk 1 above)
Comment (sayandcode) [reply programs/degen-pools/src/pools.rs:93]: Yes initially it was written like that, but there was a lifetime issue. I'm following [this stack overflow answer](https://solana.stackexchange.com/a/8450/24881), and annotating that the `remaining_accounts` has at least the same lifetime as our `FundPool` Instruction Struct accounts

File: programs/degen-pools/src/pools.rs (hunk 3 above)
Comment (sayandcode) [reply programs/degen-pools/src/pools.rs:110]: Will pursue

File: programs/degen-pools/src/pools.rs (hunk 4 above)
Comment (sayandcode) [reply programs/degen-pools/src/pools.rs:113]: When writing the tests, I noticed the option accounts weren't being written to. I couldn't find any normal sources, but Claude was able to help. Tests seem to corroborate that it works. Is there a better way to do the same?

File: programs/degen-pools/src/pools.rs (hunk 2 above)
Comment (sayandcode) [reply programs/degen-pools/src/pools.rs:117]: Checking now. Thanks for telling, I missed this check

**Comment by sayandcode** [comment]
PR was closed as we are able to achieve the same functionality in web3js by simply bundling multiple `enter_pool` instructions into a single transaction

**Pull request closed without merging**